src/Http/Router.php

```diff
@@ -45,44 +36,32 @@ public function __construct(RouteParser $routeParser, DataGenerator $dataGenerat
      */
     public function addRoute(Route $route): void
     {
-        if (!$route instanceof Route) {
-            throw RouterException::invalidRoute($route);
-        }
-        $routeDatas = $this->routeParser->parse($route->getExpression());
-        foreach ((array) $route->getMethod() as $method) {
-            foreach ($routeDatas as $routeData) {
-                $this->dataGenerator->addRoute($method, $routeData, $route);
-            }
-        }
+        $baseRoute = $route->getBaseRoute();
+        $this->routeCollection->add($route->getName(), $baseRoute);
+        $this->routes[$route->getName()] = $route;
     }
 
     /**
      * Finds route matching clients request.
      *
      * @param string $method request method.
-     * @param string $uri request uri.
+     * @param string $path request path.
      * @return Route
      */
-    public function findRoute(string $method, string $uri): Route
+    public function findRoute(string $method, string $path): Route
     {
-        $dispatcher = new GroupCountBased($this->getData());
-        $info = $dispatcher->dispatch($method, $uri);
-
-        switch ($info[0]) {
-            case Dispatcher::NOT_FOUND:
-                throw NotFoundException::notFound($uri, $method);
-
-            case Dispatcher::METHOD_NOT_ALLOWED:
-                $allowedMethods = $info[1];
-                throw MethodNotAllowedException::methodNotAllowed($uri, $method, $allowedMethods);
-
-            case Dispatcher::FOUND:
-                return $info[1]->withAttributes($info[2]);
+        $matcher = new UrlMatcher($this->routeCollection, new RequestContext('/', $method));
+        try {
+            $route = $matcher->match($path);
+        } catch (ResourceNotFoundException $exception) {
+            throw NotFoundException::notFound($path, $method);
```

Comment:
You should always use `$previous` when chaining exceptions.

Reply:
Well, I would but in this case scenario you would expose other framework's code and the stack will give you actually no valueable information.

Reply:
That's normally true in cases like that, however I do that as a rule of a thumb - just because it's 3rd party code, it doesn't mean it's pristine and bug free. One day something can break while there's nothing wrong in your code (potentially even when there was no change in your code, but a regression in a 3rd party).